contributingGuides/review/RULES.md

```diff
@@ -0,0 +1,203 @@
+```
+
+Bad:
+```tsx
+const reportIdFromRoute = report.reportId;
+const reportExists = Object.values(reports).includes(reportIdFromRoute);
+```
+
+---
+
+### [PERF-3] Use early returns in array iteration methods
+- **Severity**: SHOULD
+- **Condition**: When using `.every()`, `.some()`, or similar methods, perform simple checks first with early returns before expensive operations.
+
+Good:
+```tsx
+const areAllTransactionsValid = transactions.every((transaction) => {
+    if (!transaction.rawData || transaction.amount <= 0) return false;
+    const validation = validateTransaction(transaction);
+    return validation.isValid;
+});
+```
+
+Bad:
+```tsx
+const areAllTransactionsValid = transactions.every((transaction) => {
+    const validation = validateTransaction(transaction);
+    return validation.isValid;
+});
+```
+
+---
+
+### [PERF-4] Use OnyxListItemProvider hooks instead of useOnyx in renderItem
+- **Severity**: MUST
+- **Condition**: Components rendered inside `renderItem` functions should use dedicated hooks from `OnyxListItemProvider` instead of individual `useOnyx` calls.
+
+Good:
+```tsx
+// ReportActionItem.tsx
+const personalDetails = usePersonalDetails();
+```
+
+Bad:
+```tsx
+// ReportActionItem.tsx
+const [personalDetails] = useOnyx(ONYXKEYS.PERSONAL_DETAILS_LIST);
+```
+
+---
+
+### [PERF-5] Memoize objects and functions passed as props
+- **Severity**: MUST
+- **Condition**: Objects and functions passed as props should be properly memoized or simplified to primitive values to prevent unnecessary re-renders.
+
+Good:
+```tsx
+const reportData = useMemo(() => ({
+    reportID: report.reportID,
+    type: report.type,
+    isPinned: report.isPinned,
+}), [report.reportID, report.type, report.isPinned]);
+
+return <ReportActionItem report={reportData} />
+```
+
+Bad:
+```tsx
+const [report] = useOnyx(`ONYXKEYS.COLLECTION.REPORT${iouReport.id}`);
+
+return <ReportActionItem report={report} />
+```
+
+---
+
+### [PERF-6] Avoid inline object/array/function creation in JSX
+- **Severity**: SHOULD
+- **Condition**: Objects, arrays, and functions should not be created inline in JSX. Use `useMemo` or `useCallback` to memoize them when there's actual performance benefit.
+
+Good:
+```tsx
+const reportActionItemStyle = useMemo(() => [styles.container, styles.flex], []);
+const handleSelect = useCallback(() => {
+    onSelectRow(item);
+}, [onSelectRow, item]);
+
+<ReportActionItem
+    style={reportActionItemStyle}
+    onSelect={handleSelect}
+    reportID={report.reportID}
+    isPinned={report.isPinned}
+/>
+```
+
+Bad:
+```tsx
+<ReportActionItem
+    style={[styles.container, styles.flex]}
+    onSelect={() => onSelectRow(item)}
+    report={report}
+/>
+```
+
+---
+
+### [PERF-7] Use shallow comparisons instead of deep comparisons
+- **Severity**: SHOULD
+- **Condition**: In `React.memo` and similar optimization functions, compare only specific relevant properties instead of using deep equality checks.
```

Comment:
Unless you need to compare the whole object, right?

Reply:
In general any custom props comparison function is considered a code smell (that should be resolved in a different component structure/props being passed), because it only surfaces that the team saw a performance bottleneck caused by too many rerenders and decided to put a stop to some of them, and that later grew in size.
I'd say a deepEqual is only making the above (which is already bad) even worse because this will run before every render, making it that more expensive.

Reply:
Additionally, it's very rare that you 'have to' pass a whole object. Likely you can just pass a few primitive types, and this is going to have the best performance characteristic while being super easy to understand (with no custom comparator function being required in the future).

Reply:
Yeah, but passing the whole object has its benefits, like being able to modify the logic of a component based on some property of the object without having to refactor all the component usages to add a new prop.